### ISSUE TITLE
Serial output can freeze the emulator

### ISSUE THREAD

**Issue by pbfy0** [nspire-emus/firebird#84]
Large amounts of serial output causes the emulator to slow to a crawl. I'm not sure what the best way to fix this is, but maybe not sending serial output to the GUI thread one character at a time would help.

**Comment by MattWaltz** [comment]
The way I fixed this in CEmu was having a circular buffer inside the emu thread. During gui_do_stuff or the relevant function in the emu thread; a null character would be tacked onto the end of the buffer and sent using the console_string or relevant function.

**Comment by Vogtinator** [comment]
I noticed this issue as well, especially noticable when having something produce aborts recursively...

I tried some of @MattWaltz's tricks already, but while they helped a bit, they couldn't prevent the final freeze of the application, unfortunately. I guess an intermediate buffer + rate limiting would work.
I never heard about anything ```\n``` related though, what's up with that?

**Comment by MattWaltz** [comment]
Well I just mean that since Qt freezes because the line becomes too long; perhaps set a limit on the max amount of characters per line. The freeze goes away then; just adding a new line character fixes the issue.

**Comment by pbfy0** [comment]
It's definitely not because of excessively long lines; my lines are only 50 characters or so, but the program is generating hundreds per second.

**Comment by parrotgeek1** [comment]
Also, can you add the ability to input on serial?

**Comment by Vogtinator** [comment]
I'll add it to the todo list, what do you need it for? File transfer using XModem is available

**Comment by parrotgeek1** [comment]
u-boot

Also, whatever happened to CAS+ emulation? It's in the code but there is no
way to pick that model from the GUI.

**Comment by Vogtinator** [comment]
> u-boot

Oh, have fun... Porting that won't be easy, there is no support for the flash yet and while it can use foreign APIs to open files, it won't work directly in this case (I forgot the main reason though).

> Also, whatever happened to CAS+ emulation? It's in the code but there is no
way to pick that model from the GUI.

While CAS+ emulation is not entirely gone, some snippets didn't survive the refactoring, so it's practically dead.

**Comment by parrotgeek1** [comment]
There was a leaked diags image of u-boot 1.1.2! Signed with, incredibly,
production keys. It's on ti planet downloads

**Comment by Vogtinator** [comment]
Yes, that one is written by TI, used to launch early image builds.

**Comment by parrotgeek1** [comment]
You might be thinking of the cas+ u-boot "boot1". This one is for the
production Nspire!

**Comment by Vogtinator** [comment]
Yes, but only the "Nspire", not the "Nspire CAS" or even CX models.

**Comment by parrotgeek1** [comment]
It worked on my cas. Because boot2 and diags are the same for CAS and non CAS

**Comment by Vogtinator** [comment]
On http://blog.pankajp.com/2013/07/qplaintextedit-with-setmaximumblockcoun.html the author discusses how he works around the abysmal QTextEdit performance by using an internal string buffer and only appending the buffer after a flexible interval (100ms) after the last append finished processing.

Simpy increasing the maximum block count is IMO only delaying the inevitable, the freeze just happens an insignificant amount of time later.

**Comment by MattWaltz** [comment]
This is precisely why I said this forever ago :P 

> The way I fixed this in CEmu was having a circular buffer inside the emu thread (Which skipped input of any null characters!!). During gui_do_stuff or the relevant function in the emu thread; a null character would be tacked onto the end of the buffer and sent using the console_string or similar function

**Comment by pbfy0** [comment]
Only appending after a fixed interval will result in the screen never updating if output keeps coming. There would have to also be a maximum time or amount of characters which would force an append.

**Comment by MattWaltz** [comment]
> During gui_do_stuff or the relevant function in the emu thread

I mean I have already given you everything you need for this I may as well just implement it myself
